File: c/io.c
#define _POSIX_C_SOURCE 200809L
// Effectful primitives. Ordinary values (Val::Prim) that enter a program only
// as the `io` argument the CLI passes to `main`. Mirrors src/io.rs. Every user
// function stays pure; try/else rolls back values, not external effects.
#include "snel.h"
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

static const char *PRIMS[] = {"read",   "write", "args",   "env",   "exe",   "spawn",
                              "list",   "stat",  "exists", "mkdir", "rmdir", "unlink",
                              "rename", "link",  "time",   "sleep", NULL};
static int g_argc;
static char **g_argv;
void io_set_args(int argc, char **argv) {
    g_argc = argc;
    g_argv = argv;
}

Val io_tab(void) {
    Tab *t = tab_new();
    for (int i = 0; PRIMS[i]; i++) {
        Bin k = bin_str(PRIMS[i]);
        Val v;
        v.k = V_PRIM;
        v.u.bin = bin_str(PRIMS[i]);
        tab_bind(t, k, v, NULL);
    }
    Val v;
    v.k = V_TAB;
    v.u.tab = t;
    return v;
}

static Val fail(bool *ok, char *err, size_t n, const char *m) {
    snprintf(err, n, "%s", m);
    *ok = false;
    return vnil();
}

bool is_u8_col(const Col *c);
uint8_t *col_bytes(const Col *c, size_t *len);
bool col_from_vals(Val *, size_t, Col **, SnelErr *);

// a [u8] argument's bytes; NULL if not a string
static uint8_t *arg_bytes(Val *args, size_t na, size_t i, size_t *n) {
    if (i >= na || args[i].k != V_VEC || !is_u8_col(args[i].u.vec))
        return NULL;
    return col_bytes(args[i].u.vec, n);
}
// a [u8] argument as a NUL-terminated C string (caller frees)
static char *arg_cstr(Val *args, size_t na, size_t i) {
    size_t n;
    uint8_t *b = arg_bytes(args, na, i, &n);
    if (!b)
        return NULL;
    char *s = malloc(n + 1);
    memcpy(s, b, n);
    s[n] = 0;
    free(b);
    return s;
}
static Val str_val(const uint8_t *b, size_t n) {
    Val v;
    v.k = V_VEC;
    v.u.vec = col_u8s(b, n);
    return v;
}

static Val spawn(Val *args, size_t na, bool *ok, char *err, size_t errlen) {
    char *prog = arg_cstr(args, na, 0);
    if (!prog || na < 4 || args[1].k != V_VEC || args[3].k != V_VEC) {
        free(prog);
        return fail(ok, err, errlen, "spawn: (prog, [argv], stdin, [envs])");
    }
    Col *ac = args[1].u.vec;
    size_t stdin_n;
    uint8_t *stdin_b = arg_bytes(args, na, 2, &stdin_n);
    if (!stdin_b) {
        free(prog);
        return fail(ok, err, errlen, "spawn: stdin must be a string");
    }
    char **argv = calloc(ac->len + 2, sizeof(char *));
    argv[0] = prog;
    for (size_t i = 0; i < ac->len; i++) {
        Val e = col_elem(ac, i);
        Val a[1] = {e};
        argv[i + 1] = arg_cstr(a, 1, 0);
        if (!argv[i + 1]) {
            argv[i + 1] = malloc(1);
            argv[i + 1][0] = 0;
        }
        val_drop(e);
    }
    // env entries, each "KEY=VALUE", set on top of the inherited environment
    Col *ec = args[3].u.vec;
    size_t nenv = ec->len;
    char **envs = calloc(nenv + 1, sizeof(char *));
    for (size_t i = 0; i < nenv; i++) {
        Val e = col_elem(ec, i);
        Val a[1] = {e};
        char *s = arg_cstr(a, 1, 0);
        envs[i] = s ? s : calloc(1, 1);
        val_drop(e);
    }
    int in[2], out[2];
    if (pipe(in) || pipe(out)) {
        free(stdin_b);
        return fail(ok, err, errlen, "spawn: pipe");
    }
    pid_t pid = fork();
    if (pid == 0) {
        for (size_t i = 0; i < nenv; i++) {
            char *eq = strchr(envs[i], '=');
            if (eq) {
                *eq = 0;
                setenv(envs[i], eq + 1, 1);
            }
        }
        dup2(in[0], 0);
        dup2(out[1], 1);
        close(in[1]);
        close(out[0]);
        execvp(prog, argv);
        _exit(127);
    }
    close(in[0]);
    close(out[1]);
    if (write(in[1], stdin_b, stdin_n) < 0) { /* ignore */
    }
    close(in[1]);
    free(stdin_b);
    uint8_t *buf = NULL;
    size_t cap = 0, len = 0;
    uint8_t chunk[4096];
    ssize_t r;
    while ((r = read(out[0], chunk, sizeof chunk)) > 0) {
        if (len + r > cap) {
            cap = (len + r) * 2 + 16;
            buf = realloc(buf, cap);
        }
        memcpy(buf + len, chunk, r);
        len += r;
    }
    close(out[0]);
    waitpid(pid, NULL, 0);
    Val v = str_val(buf, len);
    free(buf);
    return v;
}

Val call_prim(const Bin *name, Val *args, size_t na, bool *ok, char *err, size_t errlen) {
    *ok = true;
    const uint8_t *nm = bin_bytes(name);
    size_t nl = name->len;
    if (nl == 4 && !memcmp(nm, "read", 4)) {
        char *path = arg_cstr(args, na, 0);
        if (!path)
            return fail(ok, err, errlen, "read: arg 0 must be a string ([u8])");
        FILE *f = fopen(path, "rb");
        free(path);
        if (!f)
            return fail(ok, err, errlen, "read: cannot open");
        fseek(f, 0, SEEK_END);
        long sz = ftell(f);
        fseek(f, 0, SEEK_SET);
        uint8_t *buf = malloc(sz > 0 ? sz : 1);
        fread(buf, 1, sz, f);
        fclose(f);
        Val v = str_val(buf, sz);
        free(buf);
        return v;
    }
    if (nl == 5 && !memcmp(nm, "write", 5)) {
        char *path = arg_cstr(args, na, 0);
        size_t dn;
        uint8_t *db = arg_bytes(args, na, 1, &dn);
        if (!path || !db) {
            free(path);
            free(db);
            return fail(ok, err, errlen, "write: (path, bytes) as strings");
        }
        FILE *f = fopen(path, "wb");
        free(path);
        if (!f) {
            free(db);
            return fail(ok, err, errlen, "write: cannot open");
        }
        fwrite(db, 1, dn, f);
        fclose(f);
        free(db);
        return vnil();
    }
    if (nl == 4 && !memcmp(nm, "args", 4)) {
        Val *items = malloc((g_argc ? g_argc : 1) * sizeof(Val));
        size_t n = 0;
        for (int i = 1; i < g_argc; i++)
            items[n++] = str_val((const uint8_t *)g_argv[i], strlen(g_argv[i]));
        Col *c;
        SnelErr e2;
        if (!col_from_vals(items, n, &c, &e2))
            return fail(ok, err, errlen, "args");
        Val v;
        v.k = V_VEC;
        v.u.vec = c;
        return v;
    }
    if (nl == 3 && !memcmp(nm, "env", 3)) {
        extern char **environ;
        Tab *t = tab_new();
        for (char **e = environ; *e; e++) {
            char *eq = strchr(*e, '=');
            if (!eq)
                continue;
            Bin k = bin_new((const uint8_t *)*e, (size_t)(eq - *e));
            tab_bind(t, k, str_val((const uint8_t *)(eq + 1), strlen(eq + 1)), NULL);
        }
        Val v;
        v.k = V_TAB;
        v.u.tab = t;
        return v;
    }
    if (nl == 3 && !memcmp(nm, "exe", 3)) {
        char buf[4096];
        ssize_t k = readlink("/proc/self/exe", buf, sizeof buf - 1);
        if (k < 0)
            return fail(ok, err, errlen, "exe: cannot read /proc/self/exe");
        return str_val((const uint8_t *)buf, (size_t)k);
    }
    if (nl == 5 && !memcmp(nm, "spawn", 5))
        return spawn(args, na, ok, err, errlen);
    if (nl == 4 && !memcmp(nm, "list", 4)) {
        char *path = arg_cstr(args, na, 0);
        if (!path)
            return fail(ok, err, errlen, "list: path must be a string");
        DIR *d = opendir(path);
        free(path);
        if (!d)
            return fail(ok, err, errlen, "list: cannot open directory");
        Val *items = NULL;
        size_t n = 0, cap = 0;
        struct dirent *ent;
        while ((ent = readdir(d))) {
            if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, ".."))
                continue; // match Rust read_dir
            if (n == cap) {
                cap = cap ? cap * 2 : 8;
                items = realloc(items, cap * sizeof(Val));
            }
            items[n++] = str_val((const uint8_t *)ent->d_name, strlen(ent->d_name));
        }
        closedir(d);
        Col *c;
        SnelErr e2;
        if (!col_from_vals(items, n, &c, &e2))
            return fail(ok, err, errlen, "list");
        Val v;
        v.k = V_VEC;
        v.u.vec = c;
        return v;
    }
    if (nl == 4 && !memcmp(nm, "stat", 4)) {
        char *path = arg_cstr(args, na, 0);
        if (!path)
            return fail(ok, err, errlen, "stat: path must be a string");
        struct stat st;
        int r = stat(path, &st);
        free(path);
        if (r != 0)
            return fail(ok, err, errlen, "stat: cannot stat");
        Tab *t = tab_new();
        tab_bind(t, bin_str("size"), vi64((int64_t)st.st_size), NULL);
        tab_bind(t, bin_str("mtime"), vi64((int64_t)st.st_mtime), NULL);
        tab_bind(t, bin_str("isdir"), vbit(S_ISDIR(st.st_mode)), NULL);
        Val v;
        v.k = V_TAB;
        v.u.tab = t;
        return v;
    }
    if (nl == 6 && !memcmp(nm, "exists", 6)) {
        char *path = arg_cstr(args, na, 0);
        if (!path)
            return fail(ok, err, errlen, "exists: path must be a string");
        bool e = access(path, F_OK) == 0;
        free(path);
        return vbit(e);
    }
    if (nl == 5 && !memcmp(nm, "mkdir", 5)) {
        char *p = arg_cstr(args, na, 0);
        if (!p)
            return fail(ok, err, errlen, "mkdir: path");
        int r = mkdir(p, 0777);
        free(p);
        if (r)
            return fail(ok, err, errlen, "mkdir failed");
        return vnil();
    }
    if (nl == 5 && !memcmp(nm, "rmdir", 5)) {
        char *p = arg_cstr(args, na, 0);
        if (!p)
            return fail(ok, err, errlen, "rmdir: path");
        int r = rmdir(p);
        free(p);
        if (r)
            return fail(ok, err, errlen, "rmdir failed");
        return vnil();
    }
    if (nl == 6 && !memcmp(nm, "unlink", 6)) {
        char *p = arg_cstr(args, na, 0);
        if (!p)
            return fail(ok, err, errlen, "unlink: path");
        int r = unlink(p);
        free(p);
        if (r)
            return fail(ok, err, errlen, "unlink failed");
        return vnil();
    }
    if (nl == 6 && !memcmp(nm, "rename", 6)) {
        char *a = arg_cstr(args, na, 0), *b = arg_cstr(args, na, 1);
        if (!a || !b) {
            free(a);
            free(b);
            return fail(ok, err, errlen, "rename: (from, to)");
        }
        int r = rename(a, b);
        free(a);
        free(b);
        if (r)
            return fail(ok, err, errlen, "rename failed");
        return vnil();
    }
    if (nl == 4 && !memcmp(nm, "link", 4)) {
        char *a = arg_cstr(args, na, 0), *b = arg_cstr(args, na, 1);
        if (!a || !b) {
            free(a);
            free(b);
            return fail(ok, err, errlen, "link: (from, to)");
        }
        int r = link(a, b);
        free(a);
        free(b);
        if (r)
            return fail(ok, err, errlen, "link failed");
        return vnil();
    }
    if (nl == 4 && !memcmp(nm, "time", 4)) {
        struct timespec ts;
        clock_gettime(CLOCK_REALTIME, &ts);
        return vi64((int64_t)ts.tv_sec * 1000 + ts.tv_nsec / 1000000);
    }
    if (nl == 5 && !memcmp(nm, "sleep", 5)) {
        if (na < 1 || args[0].k != V_I64)
            return fail(ok, err, errlen, "sleep needs an i64");
        int64_t ms = args[0].u.i;
        if (ms > 0) {
            struct timespec ts;
            ts.tv_sec = ms / 1000;
            ts.tv_nsec = (ms % 1000) * 1000000;
            nanosleep(&ts, NULL);
        }
        return vnil();
    }
    return fail(ok, err, errlen, "unknown primitive");
}

File: src/io.rs
// Effectful interpreter primitives. They are ordinary values (Val::Prim) that
// enter a program only as the `io` argument the CLI passes to `main` — code not
// handed `io` cannot perform effects (capability style). Every user function
// stays pure; try/else rolls back values, not external effects.

use crate::ast::Ty;
use crate::value::{Bytes, Col, Tab, Val};
use std::io::{Read, Write};
use std::rc::Rc;

// name -> (arg types, return type). Strings are [u8]; argv is [[u8]].
pub const PRIMS: &[(&str, &[PTy], PTy)] = &[
    ("read", &[PTy::Str], PTy::Str),                    // path -> bytes
    ("write", &[PTy::Str, PTy::Str], PTy::Nil),         // path, bytes -> nil
    ("args", &[], PTy::Strs),                           // -> [[u8]] argv
    ("env", &[], PTy::Env),                             // -> { NAME = value, … } (environ)
    ("exe", &[], PTy::Str),                             // -> path of the running interpreter
    // prog, argv, stdin, envs ("KEY=VALUE" each) -> stdout
    ("spawn", &[PTy::Str, PTy::Strs, PTy::Str, PTy::Strs], PTy::Str),
    // filesystem
    ("list", &[PTy::Str], PTy::Strs),                   // dir -> entry names
    ("stat", &[PTy::Str], PTy::Stat),                   // path -> { size, mtime, isdir }
    ("exists", &[PTy::Str], PTy::Bit),                  // path -> present?
    ("mkdir", &[PTy::Str], PTy::Nil),
    ("rmdir", &[PTy::Str], PTy::Nil),
    ("unlink", &[PTy::Str], PTy::Nil),
    ("rename", &[PTy::Str, PTy::Str], PTy::Nil),
    ("link", &[PTy::Str, PTy::Str], PTy::Nil),          // hard link
    // clock
    ("time", &[], PTy::I64),                            // -> milliseconds since the epoch
    ("sleep", &[PTy::I64], PTy::Nil),                   // sleep milliseconds
];

#[derive(Clone, Copy)]
pub enum PTy {
    Nil,
    Bit,
    I64,
    Str,  // a [u8]
    Strs, // a [[u8]]
    Env,  // a tab of env vars, keys unknown statically -> {}
    Stat, // { size: i64, mtime: i64, isdir: bit }
}

fn pty(t: PTy) -> Ty {
    let u8s = || Ty::Vec(Box::new(Ty::U8));
    match t {
        PTy::Nil => Ty::Nil,
        PTy::Bit => Ty::Bit,
        PTy::I64 => Ty::I64,
        PTy::Str => u8s(),
        PTy::Strs => Ty::Vec(Box::new(u8s())),
        PTy::Env => Ty::Tab(vec![]),
        PTy::Stat => Ty::Tab(vec![
            (Bytes::str("size"), Ty::I64),
            (Bytes::str("mtime"), Ty::I64),
            (Bytes::str("isdir"), Ty::Bit),
        ]),
    }
}

pub fn sig(name: &[u8]) -> Option<(Vec<Ty>, Ty)> {
    PRIMS
        .iter()
        .find(|(n, _, _)| n.as_bytes() == name)
        .map(|(_, a, r)| (a.iter().map(|t| pty(*t)).collect(), pty(*r)))
}

// The `io` tab: one Prim per primitive.
pub fn io_tab() -> Val {
    let mut t = Tab::default();
    for (n, _, _) in PRIMS {
        t.bind(Bytes::str(n), Val::Prim(Bytes::str(n)), None);
    }
    Val::Tab(Rc::new(t))
}

fn str_val(b: &[u8]) -> Val {
    Val::Vec(Rc::new(Col::u8s(b)))
}

// a [[u8]] column: each element is itself a [u8] column
fn strs_val(items: Vec<Vec<u8>>) -> Val {
    let elems: Vec<Val> = items.iter().map(|b| Val::Vec(Rc::new(Col::u8s(b)))).collect();
    Val::Vec(Rc::new(crate::ops::col_from_vals(elems).expect("homogeneous")))
}

pub fn call(name: &Bytes, args: Vec<Val>) -> Result<Val, String> {
    match name.bytes() {
        b"read" => {
            let path = as_string(&args, 0)?;
            std::fs::read(&path).map(|b| str_val(&b)).map_err(|e| e.to_string())
        }
        b"write" => {
            let path = as_string(&args, 0)?;
            let data = as_bytes(&args, 1)?;
            std::fs::write(&path, data).map(|_| Val::Nil).map_err(|e| e.to_string())
        }
        b"args" => Ok(strs_val(
            std::env::args().skip(1).map(|a| a.into_bytes()).collect(),
        )),
        b"env" => {
            let mut t = Tab::default();
            for (k, v) in std::env::vars_os() {
                t.bind(Bytes::new(k.to_string_lossy().as_bytes()), str_val(v.to_string_lossy().as_bytes()), None);
            }
            Ok(Val::Tab(Rc::new(t)))
        }
        b"exe" => std::env::current_exe()
            .map(|p| str_val(p.to_string_lossy().as_bytes()))
            .map_err(|e| e.to_string()),
        b"spawn" => spawn(&args),
        b"list" => {
            let path = as_string(&args, 0)?;
            let mut names = Vec::new();
            for e in std::fs::read_dir(&path).map_err(|e| e.to_string())? {
                names.push(e.map_err(|e| e.to_string())?.file_name().to_string_lossy().into_owned().into_bytes());
            }
            Ok(strs_val(names))
        }
        b"stat" => {
            let m = std::fs::metadata(as_string(&args, 0)?).map_err(|e| e.to_string())?;
            let mtime = m.modified().ok()
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map_or(0, |d| d.as_secs() as i64);
            let mut t = Tab::default();
            t.bind(Bytes::str("size"), Val::I64(m.len() as i64), None);
            t.bind(Bytes::str("mtime"), Val::I64(mtime), None);
            t.bind(Bytes::str("isdir"), Val::Bit(m.is_dir()), None);
            Ok(Val::Tab(Rc::new(t)))
        }
        b"exists" => Ok(Val::Bit(std::path::Path::new(&as_string(&args, 0)?).exists())),
        b"mkdir" => nilify(std::fs::create_dir(as_string(&args, 0)?)),
        b"rmdir" => nilify(std::fs::remove_dir(as_string(&args, 0)?)),
        b"unlink" => nilify(std::fs::remove_file(as_string(&args, 0)?)),
        b"rename" => nilify(std::fs::rename(as_string(&args, 0)?, as_string(&args, 1)?)),
        b"link" => nilify(std::fs::hard_link(as_string(&args, 0)?, as_string(&args, 1)?)),
        b"time" => Ok(Val::I64(
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map_or(0, |d| d.as_millis() as i64),
        )),
        b"sleep" => {
            match args.first() {
                Some(Val::I64(ms)) if *ms > 0 => std::thread::sleep(std::time::Duration::from_millis(*ms as u64)),
                Some(Val::I64(_)) => {}
                _ => return Err("sleep needs an i64".into()),
            }
            Ok(Val::Nil)
        }
        _ => Err(format!("unknown primitive io.{:?}", name)),
    }
}

fn nilify(r: std::io::Result<()>) -> Result<Val, String> {
    r.map(|_| Val::Nil).map_err(|e| e.to_string())
}

// a [u8] argument as raw bytes
fn as_bytes(args: &[Val], i: usize) -> Result<Vec<u8>, String> {
    match args.get(i) {
        Some(Val::Vec(c)) if crate::ops::is_u8_col(c) => Ok(crate::ops::col_bytes(c)),
        _ => Err(format!("arg {} must be a string ([u8])", i)),
    }
}
fn as_string(args: &[Val], i: usize) -> Result<String, String> {
    Ok(String::from_utf8_lossy(&as_bytes(args, i)?).into_owned())
}

// a [[u8]] argument as a vector of byte strings
fn as_strings(args: &[Val], i: usize) -> Result<Vec<Vec<u8>>, String> {
    match args.get(i) {
        Some(Val::Vec(c)) => (0..c.len)
            .map(|j| match crate::print::col_elem(c, j) {
                Val::Vec(s) if crate::ops::is_u8_col(&s) => Ok(crate::ops::col_bytes(&s)),
                _ => Err(format!("arg {i} must be a [[u8]] of strings")),
            })
            .collect(),
        _ => Err(format!("arg {i} must be a [[u8]]")),
    }
}

fn spawn(args: &[Val]) -> Result<Val, String> {
    let prog = as_string(args, 0)?;
    let argv = as_strings(args, 1)?;
    let stdin = as_bytes(args, 2)?;
    use std::process::{Command, Stdio};
    let mut cmd = Command::new(&prog);
    cmd.args(argv.iter().map(|a| String::from_utf8_lossy(a).into_owned()))
        .stdin(Stdio::piped())
        .stdout(Stdio::piped());
    // each env entry is "KEY=VALUE"; added on top of the inherited environment
    for entry in as_strings(args, 3)? {
        if let Some(eq) = entry.iter().position(|&b| b == b'=') {
            cmd.env(
                String::from_utf8_lossy(&entry[..eq]).into_owned(),
                String::from_utf8_lossy(&entry[eq + 1..]).into_owned(),
            );
        }
    }
    let mut child = cmd.spawn().map_err(|e| e.to_string())?;
    child.stdin.take().unwrap().write_all(&stdin).map_err(|e| e.to_string())?;
    let mut out = Vec::new();
    child.stdout.take().unwrap().read_to_end(&mut out).map_err(|e| e.to_string())?;
    child.wait().map_err(|e| e.to_string())?;
    Ok(str_val(&out))
}
